cpp/src/strings/copying/concatenate.cu

```diff
@@ -31,93 +35,308 @@ namespace strings
-    auto d_views = device_views.data().get();
-    // compute size of the output chars column
-    size_t total_bytes = thrust::transform_reduce( execpol->on(stream),
-        d_views, d_views + device_views.size(),
-        [] __device__ (auto d_view) {
-            if( d_view.size()==0 )
-                return static_cast<size_t>(0);
-            auto d_offsets = d_view.child(strings_column_view::offsets_column_index).template data<int32_t>();
-            size_t size = d_offsets[d_view.size()+d_view.offset()] - d_offsets[d_view.offset()];
-            return size;
-        }, static_cast<size_t>(0), thrust::plus<size_t>());
-    CUDF_EXPECTS( total_bytes < std::numeric_limits<size_type>::max(), "total size of strings is too large for cudf column" );
-
-    // create chars column
-    auto chars_column = make_numeric_column( data_type{INT8}, total_bytes, mask_state::UNALLOCATED, stream, mr);
-    auto d_new_chars = chars_column->mutable_view().data<char>();
-
-    // create offsets column
-    auto offsets_column = make_numeric_column( data_type{INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
-    auto offsets_view = offsets_column->mutable_view();
-    auto d_new_offsets = offsets_view.data<int32_t>();
+  // Compute output sizes
+  auto const device_views = create_strings_device_views(columns, stream);
+  auto const& d_views = std::get<1>(device_views);
+  auto const& d_input_offsets = std::get<2>(device_views);
+  auto const& d_partition_offsets = std::get<3>(device_views);
+  auto const strings_count = std::get<4>(device_views);
+  auto const total_bytes = std::get<5>(device_views);
+  auto const offsets_count = strings_count + 1;
+
+  if( strings_count == 0 ) {
+    return make_empty_strings_column(mr, stream);
+  }
+
+  CUDF_EXPECTS( strings_count < std::numeric_limits<size_type>::max(), 
+      "total number of strings is too large for cudf column" );
+  CUDF_EXPECTS( total_bytes < std::numeric_limits<size_type>::max(),
+      "total size of strings is too large for cudf column" );
+
+  bool const has_nulls = std::any_of(
+      columns.begin(), columns.end(),
+      [](auto const& col) { return col.has_nulls(); });
 
+  // create chars column
+  auto chars_column = make_numeric_column( data_type{INT8}, total_bytes, mask_state::UNALLOCATED, stream, mr);
+  auto d_new_chars = chars_column->mutable_view().data<char>();
+
+  // create offsets column
+  auto offsets_column = make_numeric_column( data_type{INT32}, offsets_count, mask_state::UNALLOCATED, stream, mr);
+  auto offsets_view = offsets_column->mutable_view();
+  auto d_new_offsets = offsets_view.data<int32_t>();
+
+  rmm::device_buffer null_mask;
+  size_type null_count{};
+  bitmask_type* mask_data{};
+  if (has_nulls) {
+    null_mask = create_null_mask( strings_count, mask_state::UNINITIALIZED, stream, mr );
+    mask_data = reinterpret_cast<bitmask_type*>(null_mask.data());
+  }
+
+  // From benchmark data, the fused kernel appears to perform better when the
```

Comment:
This is way cool. I feel it is kind of buried here and these values should be const values defined at the top of the file with some interesting documentation perhaps. Also, the values seem to be independent of the size of the individual columns or the size of their children?

Reply:
Now that I think about it, given that the performance seems to decrease the more the mean number of chars per string increases, I'm curious to try replacing just the chars kernel with a memcpy for each input.

Reply:
I'd rather it be the case that the fused kernel implementation were just faster in all cases instead of there being a tradeoff depending on the nature of the inputs! Having both implementations in place makes it quite a bit more complicated...

Reply:
Replacing the chars kernel sounds like a good idea.

Reply:
Unfortunately, using memcpy for the chars column is only more performant than the kernel when the strings are very large.

However, the kernel for the offsets column is more performant in all cases, so it occurs to me that I might as well always use it, and use the string sizes heuristic only to branch between how the chars column is handled. This further increased the performance for large strings, but I'll need to adjust the point where the heuristic switches to memcpy.